commands/fetch_templates.go

```diff
@@ -18,113 +19,179 @@ import (
+	if err != nil {
+		removeArchive(archive)
+		return err
+	}
 
-	err := fetchMasterZip(templateUrl)
+	log.Printf("Attempting to expand templates from %s\n", archive)
 
-	zipFile, err := zip.OpenReader(ZipFileName)
+	preExistingLanguages, fetchedLanguages, err := expandTemplatesFromZip(archive, overwrite)
 	if err != nil {
 		return err
 	}
 
-	log.Printf("Attempting to expand templates from %s\n", ZipFileName)
+	if len(preExistingLanguages) > 0 {
+		log.Printf("Cannot overwrite the following %d directories: %v\n", len(preExistingLanguages), preExistingLanguages)
+	}
+
+	log.Printf("Fetched %d template(s) : %v from %s\n", len(fetchedLanguages), fetchedLanguages, templateURL)
+
+	err = removeArchive(archive)
+
+	return err
+}
+
+// expandTemplatesFromZip() takes a path to an archive, and whether or not
+// we are allowed to overwrite pre-existing language templates. It returns
+// a list of languages that already exist (could not be overwritten), and
+// a list of languages that are newly downloaded.
+func expandTemplatesFromZip(archive string, overwrite bool) ([]string, []string, error) {
+	var existingLanguages []string
+	var fetchedLanguages []string
+	availableLanguages := make(map[string]bool)
+
+	zipFile, err := zip.OpenReader(archive)
+	if err != nil {
+		return nil, nil, err
+	}
 
 	for _, z := range zipFile.File {
-		relativePath := strings.Replace(z.Name, "faas-cli-master/", "", -1)
-		if strings.Index(relativePath, "template") == 0 {
-			fmt.Printf("Found \"%s\"\n", relativePath)
-			rc, err := z.Open()
-			if err != nil {
-				return err
-			}
+		var rc io.ReadCloser
+		var isDirectory bool
 
-			err = createPath(relativePath, z.Mode())
-			if err != nil {
-				return err
-			}
+		relativePath := z.Name[strings.Index(z.Name, "/")+1:]
+		if strings.Index(relativePath, "template/") != 0 {
+			// Process only directories inside "template" at root
+			continue
+		}
+
+		if pathSplit := strings.Split(relativePath, "/"); len(pathSplit) > 2 {
+			language := pathSplit[1]
 
-			// If relativePath is just a directory, then skip expanding it.
-			if len(relativePath) > 1 && relativePath[len(relativePath)-1:] != string(os.PathSeparator) {
-				err = writeFile(rc, z.UncompressedSize64, relativePath, z.Mode())
-				if err != nil {
-					return err
+			// We know that this path is a directory if the last character is a "/"
+			isDirectory = relativePath[len(relativePath)-1:] == "/"
+
+			// Check if this is the root directory for a language (at ./template/lang)
+			if len(pathSplit) == rootLanguageDirSplitCount && isDirectory {
+				// template/language/
+
+				if !canWriteLanguage(&availableLanguages, language, overwrite) {
+					existingLanguages = append(existingLanguages, language)
+					continue
+				}
+				fetchedLanguages = append(fetchedLanguages, language)
+			} else {
+				// template/language/*
+
+				if !canWriteLanguage(&availableLanguages, language, overwrite) {
+					continue
 				}
 			}
+		} else {
+			// template/
+			continue
```

Comment:
I don't like the half-dozen `continue` statements, how can we re-write this to be clearer?